persistence/playlist_repository.go

```diff
@@ -219,6 +219,18 @@ func (r *playlistRepository) refreshSmartPlaylist(pls *model.Playlist) bool {
 
 	// Re-populate playlist based on Smart Playlist criteria
 	rules := *pls.Rules
+
+	// If the playlist depends on other playlists, recursively refresh them first
+	childPlaylistIds := rules.ChildPlaylistIds()
+	for _, id := range childPlaylistIds {
+		childPls, err := r.Get(id)
+		if err != nil {
+			log.Error(r.ctx, "Error loading child playlist", "id", pls.ID, "childId", id, err)
+			return false
+		}
+		r.refreshSmartPlaylist(childPls)
```

Comment:
Definitely a fair point on performance. I did a little bit of testing that I'll include below. I could only manage to grab ~2500 songs for my test library - not sure what large libraries/playlists are in terms of how Navidrome is used but hopefully this is some useful data to go off! I used the logger to get the times from the running dev server.


**Test 1:**
A smart playlist with two playlist dependencies. Each picking a random 100 songs from a library of ~2500

| Playlist                      | Run 1 | Run 2 | Run 3 | Average |
| ----------------------------- | ----- | ----- | ----- | ------- |
| Sub-playlist A                | 8.5   | 7.2   | 5.7   | 7.1     |
| Sub-playlist B                | 7.9   | 13.4  | 6.5   | 9.3     |
| Main playlist with changes    | 27.4  | 32.9  | 22.5  | 27.6    |
| Main playlist without changes | 5.6   | 27.2  | 4.5   | 12.4    |

Difference of ~15.2ms to fetch main playlist. This aligns with the ~16.4 total milliseconds taken to refresh the sub playlists.



**Test 2:** 
A smart playlist with two playlist dependencies. Each picking a random 1000 songs from a library of ~2500

| Playlist                      | Run 1 | Run 2 | Run 3 | Average |
| ----------------------------- | ----- | ----- | ----- | ------- |
| Sub-playlist A                | 10.0  | 8.6   | 10.0    | 9.5     |
| Sub-playlist B                | 8.8   | 8.5   | 7.7   | 8.3     |
| Main playlist with changes    | 43.7  | 45    | 43    | 43.9    |
| Main playlist without changes | 25.9  | 25.3  | 23.4  | 24.9    |

Difference of ~19ms to fetch main playlist. This aligns with the ~17.8 total milliseconds taken to refresh the sub playlists.


Overall the fetching time for the sub-playlists is basically added onto the main playlist which is expected. The additional time is pretty minor (at least at this scale) for the total refresh time to the user; however it does basically double the overall time which could be an issue at a much larger scale.

An alternative may be to map these recursive refreshes into the one SQL statement directly as sub-queries (although this is probably a bit beyond what I can wrangle, at least right now).

Otherwise perhaps a boolean flag on the playlist to indicate whether it should use the 'dynamic refresh' of sub-playlists functionality.

Reply:
Can you post here the NSP files you used to run this tests? I can run it agains my 83K songs library.

Reply:
Wow thats a serious test library...
Here are my nsp files (as json files so github would allow the upload). Basically I used a bulk tagger to set half my library with a comment of A and the other half with B.

[mainPlaylist.json](https://github.com/user-attachments/files/15906337/mainPlaylist.json)
[selectionB.json](https://github.com/user-attachments/files/15906339/selectionB.json)
[selectionA.json](https://github.com/user-attachments/files/15906340/selectionA.json)